### PR TITLE
Clio 29318 focus component

### PR DESCRIPTION
**Pull request by JefCurtis**

### CODE REVIEW

File: themis_components/thDisclosure/examples/1 - simple/html.html [hunk 1]
```diff
@@ -1,6 +1,11 @@
 <html ng-app="thDemo">
-  <body style="height: 25px;" ng-controller="DemoCtrl as demo">
+  <body style="height: 225px;" ng-controller="DemoCtrl as demo">
     <th-disclosure-toggle name="unique-name" expanded="demo.expanded">Show hidden text</th-disclosure-toggle>
-    <th-disclosure-content name="unique-name">Hidden text</th-disclosure-content>
+    <th-disclosure-content name="unique-name">
```
Comment (PANCAKES) [inline themis_components/thDisclosure/examples/1 - simple/html.html]: I think we should move this example to its own example in the ```with-focus``` directive folder. You can still use the disclosure there, but you are showcasing the with-focus. Here it is out of place for sure.

Comment (JefCurtis) [inline themis_components/thDisclosure/examples/1 - simple/html.html]: Yep will do.

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee [hunk 2]
```diff
@@ -25,14 +25,20 @@ angular.module 'ThemisComponents'
...
-    controller: ($element) ->
+    controller: ($scope, $element, $transclude) ->
+
+      processContent = ->
+        $transclude (transEl) ->
+          isEmpty = $element.children().length == 0
+          $element.append transEl if isEmpty
 
       animateToggle = =>
         height = getActualHeight $element
```
Comment (PANCAKES) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: You can get rid of this calculation of height. You recalculate before you use it in the ```if @expanded``` block

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee [hunk 3]
```diff
@@ -12,9 +12,9 @@ describe "ThemisComponents: Directive: thDisclosureContent", ->
         </th-disclosure-content>
       """)
 
-    it "renders a <ng-transclude> component", ->
+    it "component is empty", ->
```
Comment (PANCAKES) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: Should we have another test if the component is not empty?

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee [hunk 4]
```diff
@@ -23,8 +23,8 @@ describe "ThemisComponents: Directive: thDisclosureContent", ->
     beforeEach ->
       {element} = compileDirective("""
         <div>
-          <th-disclosure-toggle name="unique-id">Toggle</th-disclosure-toggle>
-          <th-disclosure-content name="unique-id">Content</th-disclosure-content>
+          <th-disclosure-toggle name="unique-id-1">Toggle</th-disclosure-toggle>
```
Comment (PANCAKES) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: I am super confused why we changed the id here

File: themis_components/thWithFocus/thWithFocus.directive.coffee [hunk 5]
```diff
@@ -0,0 +1,17 @@
+angular.module("ThemisComponents")
+  .directive "withFocus", ($timeout) ->
+    restrict: "A"
+    link: (scope, element) ->
+      setFocus = (el) ->
+        # Set to 301ms to account for th-modal's CSS transition time.
```
Comment (PANCAKES) [inline themis_components/thWithFocus/thWithFocus.directive.coffee]: Always good to document these - nice

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee (hunk 2 above)
Comment (JefCurtis) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: Right. Missed that.

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee (hunk 3 above)
Comment (JefCurtis) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: When the disclosure-content is in "hidden" state the component is always empty because we have not called the transclude function and appended the content.

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee (hunk 4 above)
Comment (JefCurtis) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: Ahh that was for testing. I wanted to make sure I had the right controller. ugh.. testing sucks. Will change back.

File: themis_components/thDisclosure/index.scss [hunk 6]
```diff
@@ -2,6 +2,7 @@ th-disclosure-content {
   overflow: hidden;
   height: 0;
   display: block;
+
```
Comment (andreisoare) [inline themis_components/thDisclosure/index.scss]: Extra blank line.

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee [hunk 7]
```diff
@@ -25,14 +25,19 @@ angular.module 'ThemisComponents'
...
       name: '@'
-    template: require './thDisclosureContent.template.html'
     bindToController: true
     controllerAs: 'thDisclosureContent'
-    controller: ($element) ->
+    controller: ($scope, $element, $transclude) ->
+
+      processContent = ->
+        $transclude (transEl) ->
+          isEmpty = $element.children().length == 0
```
Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: `is 0`

File: themis_components/thWithFocus/examples/2 - In Modal/html.html [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<html ng-app="thDemo">
```
Comment (andreisoare) [inline themis_components/thWithFocus/examples/2 - In Modal/html.html]: Let's add a 3rd example with disclosure, so that we can see that it works.

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee [hunk 9]
```diff
@@ -25,14 +25,19 @@ angular.module 'ThemisComponents'
...
     scope:
       name: '@'
-    template: require './thDisclosureContent.template.html'
     bindToController: true
     controllerAs: 'thDisclosureContent'
-    controller: ($element) ->
+    controller: ($scope, $element, $transclude) ->
+
+      processContent = ->
+        $transclude (transEl) ->
```
Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: I think you need to add a listener for thDisclosureContent's $scope.$destroy and manually destroy the transcluded scope. Something like: 

```coffeescript
processContent = ->
  $transclude (transEl, transScope) ->
    isEmpty = $element.children().length == 0
    $element.append transEl if isEmpty
    $scope.on '$destroy', -> transScope.$destroy
```

Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: ^ this is coming from [here](https://docs.angularjs.org/api/ng/service/$compile#transclusion-functions):

```
It is normal practice to attach your transcluded content (clone) to the DOM inside your clone attach function:

var transcludedContent, transclusionScope;

$transclude(function(clone, scope) {
  element.append(clone);
  transcludedContent = clone;
  transclusionScope = scope;
});

Later, if you want to remove the transcluded content from your DOM then you should also destroy the associated transclusion scope:

transcludedContent.remove();
transclusionScope.$destroy();

Best Practice: if you intend to add and remove transcluded content manually in your directive (by calling the transclude function to get the DOM and calling element.remove() to remove it), then you are also responsible for calling $destroy on the transclusion scope.
```

Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: And it would be good to test this somehow, to make sure it works properly.

Comment (PANCAKES) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: See, this is where I learn something. Great catch, Andrei

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee [hunk 10]
```diff
@@ -59,7 +59,7 @@ describe "ThemisComponents: Directive: thDisclosureContent", ->
       , 301 # The animation duration is 300ms
 
     it "toggles closed and back open", ->
-      ctrl = element.find('th-disclosure-content ng-transclude').scope().thDisclosureContent
+      ctrl = element.find('th-disclosure-content span').scope().$$prevSibling.thDisclosureContent
```
Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: `$$prevSibling` is very hard to figure out what element you're referring to. Can we avoid this?

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee [hunk 11]
```diff
@@ -25,14 +25,19 @@ angular.module 'ThemisComponents'
...
+          isEmpty = $element.children().length == 0
+          $element.append transEl if isEmpty
 
       animateToggle = =>
-        height = getActualHeight $element
         if @expanded
+          processContent()
```
Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: You should try removing the content from the DOM when @expanded is false, so that you can re-insert it when @expanded is true and have `with-focus` kick in again.

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee [hunk 12]
```diff
@@ -25,14 +25,19 @@ angular.module 'ThemisComponents'
     transclude: true
     scope:
       name: '@'
-    template: require './thDisclosureContent.template.html'
     bindToController: true
     controllerAs: 'thDisclosureContent'
-    controller: ($element) ->
+    controller: ($scope, $element, $transclude) ->
+
```
Comment (andreisoare) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: I noticed very often that people leave a blank line after starting a function. I don't see any reason why we need to so this. @PANCAKES @seanhealy thoughts?

File: themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee (hunk 10 above)
Comment (JefCurtis) [inline themis_components/thDisclosure/thDisclosureContent.directive.spec.coffee]: It's confusing I know but because of how transclusion scopes seem to work they are actually siblings. The directive's scope is isolated (doesn't have any parents) and the transcluded scope is a descendant of the "thDemo" scope which makes them sibling. 
![screenshot 2016-01-15 14 23 48]
(https://cloud.githubusercontent.com/assets/3892972/12367035/fe876380-bb93-11e5-96e2-4bb756b64a1d.png)
http://teropa.info/blog/2015/06/09/transclusion.html

File: themis_components/thDisclosure/thDisclosureContent.directive.coffee (hunk 11 above)
Comment (JefCurtis) [inline themis_components/thDisclosure/thDisclosureContent.directive.coffee]: Sadly I cannot seem to get this to work. I can't seem to get access to the raw template in order to recompile it and restart the directives attached to it. I'll take any help I can get on this one though.
